common/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+pub mod types;
+
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use tokio::sync::mpsc;
+use types::*;
+
+/// TODO: Provide error types.
+///
+/// Note: This trait is quite subject to change.
+#[async_trait]
+pub trait AuthorizedNetwork {
+    /// Joins the network with an authorized identity.
+    async fn new(
+        public_key: PublicKey,
+        private_key: PrivateKey,
+        known_ids: Vec<PublicKey>,
+        known_peers: Vec<std::net::Ipv4Addr>,
+        network_id: String,
+    ) -> Result<Self, String>
+    where
+        Self: Sized;
+    /// Broadcasts a message to the network, after signed by the key given to this instance.
+    async fn broadcast(&self, message: &[u8]) -> Result<(), String>;
+    /// Creates a receiver for every message broadcasted to the network, except the one sent by this instance.
+    async fn create_recv_queue(&self) -> Result<mpsc::Receiver<Vec<u8>>, ()>;
+    /// Provides an estimated lists of live nodes that are eligible and identified by their public keys.
+    async fn get_live_list(&self) -> Result<Vec<PublicKey>, ()>;
+}
+
+/// TODO: Provide error types.
+///
+/// Note: This trait is quite subject to change.
+#[async_trait]
+pub trait UnauthorizedNetwork {
+    /// Joins the network with an authorized identity.
+    async fn new(known_peers: Vec<std::net::Ipv4Addr>, network_id: String) -> Result<Self, String>
+    where
+        Self: Sized;
+    /// Broadcasts a message to the network, after signed by the key given to this instance.
+    async fn broadcast(&self, message: &[u8]) -> Result<(), String>;
+    /// Creates a receiver for every message broadcasted to the network, except the one sent by this instance.
+    async fn create_recv_queue(&self) -> Result<mpsc::Receiver<Vec<u8>>, ()>;
+    /// Provides an estimated lists of live nodes identified by their IP addresses
+    async fn get_live_list(&self) -> Result<Vec<std::net::Ipv4Addr>, ()>;
+}
+
+/// TODO: Provide error types.
+///
+/// Note: This trait is quite subject to change.
+#[async_trait]
+pub trait KVStore {
+    /// Creates an empty store with the path to newly create.
+    async fn new(path: &str) -> Result<Self, ()>
+    where
+        Self: Sized;
+    /// Open an existing store with the path given.
+    async fn open(path: &str) -> Result<Self, ()>
+    where
+        Self: Sized;
+    /// Records the current state to the persistent storage.
+    async fn commit_checkpoint(&mut self) -> Result<(), ()>;
+    /// Reverts all the changes made since the last checkpoint.
+    async fn reset_to_latest_checkpoint(&mut self) -> Result<(), ()>;
+    /// Inserts a key-value pair into the store. If exists, it will be overwritten.
+    async fn insert(&mut self, key: Hash256, value: &[u8]) -> Result<(), ()>;
+    /// Remove a key-value pair from the store. If not exists, it will fail.
```

Comment:
Removes?

I think we should discuss about and come up with some consistent rules for documenting.

Reply:
fixed. Let's discuss the comment rule problem later in Discord